x/meshsecurity/keeper/handler_plugin.go

```diff
@@ -0,0 +1,96 @@
+package keeper
+
+import (
+	"encoding/json"
+
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/osmosis-labs/mesh-security-sdk/x/meshsecurity/contract"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+)
+
+type msKeeper interface {
+	IsAuthorized(ctx sdk.Context, contractAddr sdk.AccAddress) (bool, error)
+	Delegate(ctx sdk.Context, actor sdk.AccAddress, addr sdk.ValAddress, coin sdk.Coin) (sdk.AccAddress, error)
+	Undelegate(ctx sdk.Context, actor sdk.AccAddress, addr sdk.ValAddress, coin sdk.Coin) error
+}
+
+type CustomMsgHandler struct {
+	k msKeeper
+}
+
+func (h CustomMsgHandler) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, _ string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom == nil {
+		return nil, nil, wasmtypes.ErrUnknownMsg
+	}
+	var customMsg contract.CustomMsg
+	if err := json.Unmarshal(msg.Custom, &customMsg); err != nil {
+		return nil, nil, sdkerrors.ErrJSONUnmarshal.Wrap("custom message")
+	}
+	if customMsg.VirtualStake != nil {
+		// not our message type
+		return nil, nil, wasmtypes.ErrUnknownMsg
+	}
+
+	ok, err := h.k.IsAuthorized(ctx, contractAddr)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !ok {
+		return nil, nil, sdkerrors.ErrUnauthorized.Wrapf("contract has no permission for mesh security operations")
+	}
+
+	switch {
```

Comment:
Everything looks great here. Straight forward conversion and placing logic into the keeper